lib/tag-reader.coffee

```diff
@@ -1,44 +1,39 @@
 {Task} = require 'atom'
 ctags = require 'ctags'
-fs = require 'fs-plus'
+async = require 'async'
+getTagsFile = require "./get-tags-file"
 
 handlerPath = require.resolve './load-tags-handler'
 
 module.exports =
-  getTagsFile: ->
-    [directory] = atom.project.getDirectories()
-    return unless directory?
-
-    tagsFile = directory.resolve("tags")
-    return tagsFile if fs.isFileSync(tagsFile)
-
-    tagsFile = directory.resolve("TAGS")
-    return tagsFile if fs.isFileSync(tagsFile)
-
-    tagsFile = directory.resolve(".tags")
-    return tagsFile if fs.isFileSync(tagsFile)
-
-    tagsFile = directory.resolve(".TAGS")
-    return tagsFile if fs.isFileSync(tagsFile)
-
   find: (editor, callback) ->
     if editor.getLastCursor().getScopeDescriptor().getScopesArray().indexOf('source.ruby') isnt -1
       # Include ! and ? in word regular expression for ruby files
       range = editor.getLastCursor().getCurrentWordBufferRange(wordRegex: /[\w!?]*/g)
     else
       range = editor.getLastCursor().getCurrentWordBufferRange()
+
     symbol = editor.getTextInRange(range)
 
-    tagsFile = @getTagsFile()
+    unless symbol?.length > 0
+      return process.nextTick -> callback(null, [])
 
-    if symbol?.length > 0 and tagsFile
-      ctags.findTags(tagsFile, symbol, callback)
-    else
-      process.nextTick -> callback(null, [])
+    allTags = []
+
+    async.each(
+      atom.project.getPaths(),
+      (projectPath, done) =>
+        tagsFile = getTagsFile(projectPath)
+        return done() unless tagsFile?
+        ctags.findTags tagsFile, symbol, (err, tags=[]) ->
+          tag.directory = projectPath for tag in tags
+          allTags.push(tags...)
```

Comment:
There can be a lot of tags (100k is not unreasonable) and we've had issues in the past with using really large argument arrays like this. https://github.com/atom/underscore-plus/blob/2046d1562896761e3c38b7a647ac30b87b46ea20/src/underscore-plus.coffee#L264 is a workaround we've used in the past.

Maybe `concat` would be better here just in case?